modyn/config/schema/pipeline.py

```diff
@@ -459,6 +459,20 @@ def evaluation_transformer_function_deserialized(self) -> Callable | None:
             return deserialize_function(self.evaluation_transformer_function, "evaluation_transformer_function")
         return None
 
+    @model_validator(mode="after")
+    def can_instantiate_metric(self) -> Self:
```

Comment:
Rather than validating the `Metric.config` dictionary via a model validator, I would prefer a `MetricConfig` type that is a pydantic annotated union (like for selection strategy, trigger config, ...) of explicit metric config subtypes for every metric (e.g. `F1ScoreMetricConfig`, `AccuracyMetricConfig`, ...). I think that's fair as the metric classes are also part of our code base (and not of some library). This would make the configuration more explicit and easier to use. Also, we could strip the validation from the business logic. With the dictionaries, I faced the problem of not knowing what fields are available/necessary for certain metrics. Also, it doesn't catch the specifications of not existing options.
The model validator already helps to make things more explicit and easier to catch as validation is already done at config creation time, however, more explicit typing could verify config correctness just by typing compliance.
I understand though if you don't have the time to implement that, so I'm fine merging this as it's already an improvement over the status quo.

Reply:
i guess you're right and i took the lazy approach. will try to make this better tomorrow. thanks for the suggestion and kicking my lazy ass here :D

Reply:
As mentioned on Element, this has some implications that blow up this PR beyond its scope of allowing multiple F1 scores. Let's update this, but let's do it in a follow up PR. I had some initial implementation in https://github.com/eth-easl/modyn/pull/467/commits/acae267a9a2dcf487034a231d57ec62e58f6781e but there are some questions as mentioned on element on how to do this correctly